app/forms/openid_connect_authorize_form.rb

```diff
@@ -0,0 +1,113 @@
+class OpenidConnectAuthorizeForm
+  include ActiveModel::Model
+
+  VALID_ACR_VALUES = [
+    Saml::Idp::Constants::LOA1_AUTHN_CONTEXT_CLASSREF,
+    Saml::Idp::Constants::LOA3_AUTHN_CONTEXT_CLASSREF
+  ].freeze
+
+  attr_reader :acr_values,
+              :client_id,
+              :nonce,
+              :prompt,
+              :redirect_uri,
+              :response_type,
+              :scope,
+              :state
+
+  validates_presence_of :acr_values,
+                        :client_id,
+                        :prompt,
+                        :redirect_uri,
+                        :scope,
+                        :state
+
+  validates_inclusion_of :response_type, in: %w(code)
+  validates_inclusion_of :prompt, in: %w(select_account)
+
+  validate :validate_acr_values
+  validate :validate_client_id
+  validate :validate_redirect_uri
+  validate :validate_scope
+
+  def initialize(params)
+    @acr_values = parse_acr_values(params[:acr_values])
+    simple_attrs = %i(client_id nonce prompt redirect_uri response_type scope state)
+    simple_attrs.each do |key|
+      instance_variable_set(:"@#{key}", params[key])
+    end
+  end
+
+  def params
+    {
+      acr_values: acr_values.join(' '),
+      client_id: client_id,
+      nonce: nonce,
+      prompt: prompt,
+      redirect_uri: redirect_uri,
+      response_type: response_type,
+      scope: scope,
+      state: state
+    }
+  end
+
+  # @return [String,nil] redirect_uri or nil
+  def submit(current_user)
+    if valid?
+      code = OpenidConnectService.generate_authorization_code(
+        user: current_user, client_id: client_id, nonce: nonce
+      )
+
+      URIService.add_params(redirect_uri, code: code, state: state)
+    else
+      error_redirect_uri
+    end
+  end
+
+  private
+
+  def parse_acr_values(acr_values)
+    return [] if acr_values.blank?
+    acr_values.split(' ').compact & VALID_ACR_VALUES
+  end
+
+  def validate_acr_values
+    return if acr_values.present?
+    errors.add(:acr_values, 'no acceptable acr_values found')
+  end
+
+  def validate_client_id
+    return if service_provider.valid?
+    errors.add(:client_id, 'bad client_id')
+  end
+
+  def validate_redirect_uri
+    _uri = URI(redirect_uri)
+    sp_redirect_uri = service_provider.metadata[:redirect_uri]
+    if sp_redirect_uri.present?
+      return if sp_redirect_uri.start_with?(redirect_uri)
+      errors.add(:redirect_uri, 'redirect_uri does not match regirested redirect_uri')
+    else
+      errors.add(:redirect_uri, 'no redirect_uri for this client_id')
+    end
+  rescue ArgumentError, URI::InvalidURIError
+    errors.add(:redirect_uri, 'redirect_uri is invalid')
+  end
```

Comment:
Should these error messages be localized? And should they be explicitly tested? I did a search for these error messages and didn't find them referenced anywhere else.